### PR TITLE
feat: Introduce userroles recipe

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Introduce userroles recipe

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
 
## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/userroles/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from typing import Any, Dict, List, Union
+
+
+class AddRoleToUserOkResult:
+    def __init__(self, did_user_already_have_role: bool):
+        self.did_user_already_have_role = did_user_already_have_role
+
+
+class UnknownRoleError:
+    # self.status = "UNKNOWN_ROLE_ERROR"
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/userroles/interfaces.py:11]: done.

File: supertokens_python/recipe/userroles/recipe.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            raise_general_exception(
+                'calling testing function in non testing env')
+        UserRolesRecipe.__instance = None
+
+    @staticmethod
+    def get_instance() -> UserRolesRecipe:
+        if UserRolesRecipe.__instance is not None:
+            return UserRolesRecipe.__instance
+        raise_general_exception(
+            'Initialisation not done. Did you forget to call the SuperTokens.init function?')
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/userroles/recipe.py:87]: done.

File: supertokens_python/recipe/userroles/interfaces.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+RemoveUserRoleResult = Union[RemoveUserRoleOkResult, UnknownRoleError]
+
+
+class GetRolesForUserOkResult:
+    def __init__(self, roles: List[str]):
+        self.roles = roles
+
+
+GetRolesForUserResult = GetRolesForUserOkResult
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/userroles/interfaces.py:31]: done.

File: supertokens_python/recipe/userroles/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    def __init__(self, did_user_already_have_role: bool):
+        self.did_user_already_have_role = did_user_already_have_role
+
+
+class UnknownRoleError:
+    # self.status = "UNKNOWN_ROLE_ERROR"
+    pass
+
+
+AddRoleToUserResult = Union[AddRoleToUserOkResult, UnknownRoleError]
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/userroles/interfaces.py:15]: done.

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,36 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [unreleased]
 
 ## [0.9.0] - 2022-06-23
+### Features:
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply CHANGELOG.md:10]: fixed.

File: supertokens_python/constants.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
 # WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 # License for the specific language governing permissions and limitations
 # under the License.
-SUPPORTED_CDI_VERSIONS = ['2.9', '2.10', '2.11', '2.12', '2.13']
+SUPPORTED_CDI_VERSIONS = ['2.9', '2.10', '2.11', '2.12', '2.13', '2.14']
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/constants.py:14]: done.

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**